### PR TITLE
Mostekcm add refresh token support

### PR DESCRIPTION
**Pull request by mostekcm**
Fixed a couple of bugs and added refresh token support, this has already been pushed to 8.x-2.x-dev in the drupal module.

### CODE REVIEW

File: auth0.install [hunk 1]
```diff
@@ -57,7 +57,7 @@ function auth0_install() {
 
   $config = \Drupal::service('config.factory')->getEditable('auth0.settings');
   $config->set('auth0_login_css', $css)
-          ->set('auth0_widget_cdn', 'https://cdn.auth0.com/js/lock/10.4/lock.min.js')
+          ->set('auth0_widget_cdn', 'http://cdn.auth0.com/js/lock/11.0/lock.min.js')
```
Comment (sgmeyer) [inline auth0.install]: is this really what you want?  maybe https

File: auth0.install [hunk 2]
```diff
@@ -56,13 +56,16 @@ function auth0_install() {
...
-  $config->set('auth0_login_css', $css)
-          ->set('auth0_widget_cdn', 'https://cdn.auth0.com/js/lock/10.4/lock.min.js')
-          ->set('auth0_form_title', 'Sign In')
-          ->set('auth0_requires_verified_email', TRUE)
-          ->set('auth0_join_user_by_mail_enabled', FALSE)
+  $config->set('auth0_form_title', 'Sign In')
           ->set('auth0_allow_signup', TRUE)
+          ->set('auth0_allow_offline_access', FALSE)
           ->set('auth0_redirect_for_sso', FALSE)
+          ->set('auth0_widget_cdn', 'http://cdn.auth0.com/js/lock/10.22/lock.min.js')
```
Comment (glena) [inline auth0.install]: why dont you keep this without the protocol `//cdn...`?

File: src/Controller/AuthController.php [hunk 3]
```diff
@@ -143,77 +155,101 @@ protected function getNonce() {
...
 
-    $this->logger->error("$nonce not found in: ".implode(',', $nonces));
+    $this->logger->error("$nonce not found in: ".implode(',', $states));
     return FALSE;
   }
 
   /**
    * Build the Authorize url
    * @param $prompt none|login if prompt=none should be passed, false if not
+   * @param $returnTo local path|null if null, use default of /user
```
Comment (cocojoe) [inline src/Controller/AuthController.php]: whats the return?

File: src/Util/AuthHelper.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  const AUTH0_LOGGER = 'auth0_helper';
+  const AUTH0_DOMAIN = 'auth0_domain';
+  const AUTH0_CLIENT_ID = 'auth0_client_id';
+  const AUTH0_CLIENT_SECRET = 'auth0_client_secret';
+  const AUTH0_REDIRECT_FOR_SSO = 'auth0_redirect_for_sso';
+  const AUTH0_JWT_SIGNING_ALGORITHM = 'auth0_jwt_signature_alg';
+  const AUTH0_SECRET_ENCODED = 'auth0_secret_base64_encoded';
+  const AUTH0_OFFLINE_ACCESS = 'auth0_allow_offline_access';
+  
+  protected $eventDispatcher;
```
Comment (cocojoe) [inline src/Util/AuthHelper.php]: What does this property do in this class?

Comment (cocojoe) [reply src/Util/AuthHelper.php]: Also the other properties used in this class are not delcared?

File: src/Util/AuthHelper.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+   */
+  public function __construct() {
+    $this->logger = \Drupal::logger(AuthHelper::AUTH0_LOGGER);
+    $this->config = \Drupal::service('config.factory')->get('auth0.settings');
+    $this->domain = $this->config->get(AuthHelper::AUTH0_DOMAIN);
+    $this->client_id = $this->config->get(AuthHelper::AUTH0_CLIENT_ID);
+    $this->client_secret = $this->config->get(AuthHelper::AUTH0_CLIENT_SECRET);
+    $this->redirect_for_sso = $this->config->get(AuthHelper::AUTH0_REDIRECT_FOR_SSO);
+    $this->auth0_jwt_signature_alg = $this->config->get(AuthHelper::AUTH0_JWT_SIGNING_ALGORITHM);
+    $this->secret_base64_encoded = FALSE || $this->config->get(AuthHelper::AUTH0_SECRET_ENCODED);
```
Comment (cocojoe) [inline src/Util/AuthHelper.php]: So why do we need a boolean operator? 
What are the possible values of AUTH0_SECRET_ENCODED and AUTH0_OFFLINE_ACCESS?
I am thinking this has been added like Javascript || behaviour?

File: auth0.install (hunk 2 above)
Comment (mostekcm) [reply auth0.install]: fixing

File: src/Controller/AuthController.php [hunk 6]
```diff
@@ -42,15 +43,16 @@
  */
 class AuthController extends ControllerBase {
   const SESSION = 'auth0';
-  const NONCE = 'nonce';
```
[the remark this answers is not in the document]
Comment (mostekcm) [reply src/Controller/AuthController.php]: Just renaming to STATE for the session storage variable name.

File: src/Controller/AuthController.php (hunk 3 above)
Comment (mostekcm) [reply src/Controller/AuthController.php]: added

File: src/Util/AuthHelper.php (hunk 4 above)
Comment (mostekcm) [reply src/Util/AuthHelper.php]: fixed

File: src/Util/AuthHelper.php (hunk 5 above)
Comment (mostekcm) [reply src/Util/AuthHelper.php]: offlineAccess isn't used so I deleted it.

For the secret_base64_encoded, I'm just preserving the same line that was in the callback.

File: src/Util/AuthHelper.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    $this->offlineAccess = FALSE || $this->config->get(AuthHelper::AUTH0_OFFLINE_ACCESS);
+  }
+
+  /**
+   * @param $refreshToken the refresh token to use to get the user
+   * @throws \Drupal::auth0::Exception::RefreshTokenFailedException
+   */
```
[the remark this answers is not in the document]
Comment (mostekcm) [reply src/Util/AuthHelper.php]: fixed

**Pull request merged**